pkg/builder/webhook.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/conversion"
+)
+
+// WebhookBuilder builds a Controller.
+type WebhookBuilder struct {
+	apiType runtime.Object
+	mgr     manager.Manager
+	config  *rest.Config
+}
+
+func WebhookManagedBy(m manager.Manager) *WebhookBuilder {
+	return &WebhookBuilder{mgr: m}
+}
+
+// For defines the type of Object being *reconciled*, and configures the ControllerManagedBy to respond to create / delete /
+// update events by *reconciling the object*.
+// This is the equivalent of calling
+// Watches(&source.Kind{Type: apiType}, &handler.EnqueueRequestForObject{})
+// If the passed in object has implemented the admission.Defaulter interface, a MutatingWebhook will be wired for this type.
+// If the passed in object has implemented the admission.Validator interface, a ValidatingWebhook will be wired for this type.
+// TODO: figure out if we support only one object or multiple objects
+func (blder *WebhookBuilder) For(apiType runtime.Object) *WebhookBuilder {
+	blder.apiType = apiType
+	return blder
+}
+
+// Complete builds the Application ControllerManagedBy.
+func (blder *WebhookBuilder) Complete() error {
+	// Set the Config
+	if err := blder.doConfig(); err != nil {
+		return err
+	}
+
+	// Set the Manager
+	if err := blder.doManager(); err != nil {
+		return err
+	}
+
+	// Set the Webook if needed
+	if err := blder.doWebhook(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (blder *WebhookBuilder) doConfig() error {
+	if blder.config != nil {
+		return nil
+	}
+	if blder.mgr != nil {
+		blder.config = blder.mgr.GetConfig()
+		return nil
+	}
+	var err error
+	blder.config, err = getConfig()
+	return err
+}
+
+func (blder *WebhookBuilder) doManager() error {
```

Comment:
not needed -- was only in the old builder for legacy reasons (before we switched to `ManagedBy`)